File: js/160_IntersectionofTwoLinkedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    var pA = headA;
    var pB = headB;
    var tA = null;
    var tB = null;
    if( headA === null || headB === null ) return null;
    while(1){
        if( pA === null ) pA = headB;
        if( pB === null ) pB = headA;
        if( pA.next === null ) tA = pA;
        if( pB.next === null ) tB = pB;
        if( tA !== null && tB !== null && tA != tB) return null;
        if( pA === pB ) return pA;
        pA = pA.next;
        pB = pB.next;
    }
    return null;
};

File: cpp/160_IntersectionofTwoLinkedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *pA = headA;
        ListNode *pB = headB;
        ListNode *tA = NULL;
        ListNode *tB = NULL;
        if(headA == NULL || headB == NULL) return NULL;
        while(1){
            if(pA == NULL) pA = headB;
            if(pB == NULL) pB = headA;
            if(pA->next == NULL) tA = pA;
            if(pB->next == NULL) tB = pB;
            if(tA != NULL && tB != NULL && tA != tB) return NULL;
            if(pA == pB) return pA;
            pA = pA->next;
            pB = pB->next;
        }
        return NULL;
    }
};